### DISCUSSION TITLE
Ch 1 `view-source` Exercise Clarification

### DISCUSSION THREAD

**Discussion by mipsumlo** [browserengineering/book#559] [Q&A]
In Chapter 1's Exercise:

_view-source: In addition to HTTP and HTTPS, there are other schemes, such as view-source; navigating in a real browser to view-source:http://browser.engineering/http.html shows the HTML source of this chapter rather than its rendered output. Add support for the view-source scheme. Your browser should print the entire HTML file as if it was text. Hint: To do so, you can utilize the entities from the previous exercise, and add an extra transform() method that adjusts the input to show() when in view-source mode, like this: show(transform(body))_

the Hint is a bit confusing. Is there an expected difference of output here compared to just printing a plain text response of the raw HTML content? Not sure where encodings /`transform` would be useful given an input of raw HTML, unless the exercise is implying HTML validation as well prior to outputting the `view-source` content (in which case, that would make sense but it's unclear at this step). Inspecting the `view-source` output in my browser (Chrome on Mac OSX) does show a `<table>` rendering of the source HTML, but I believe this is a browser default (not a good use of `<table>` in my opinion...), and plain text output seems sufficient.

**Comment by pavpanchekha** [comment]
If you pass HTML source code like `<b>hi</b>` to `show` it won't show the angle brackets, just the text `hi`, and that text will be bolded. Of course, you could skip `show` entirely, but the idea is that you want to modify the network code to support `view-source` without changing the `show` function. That way, `view-source` will automatically be supported once you add a graphical mode in the next chapter.

**Reply by mipsumlo** [reply]
Ah, that makes sense; thank you for the clarification! I had refactored my code with my own HTML parsing module which `show` calls, but my current implementation skips `show` if `view-source` is present to avoid unnecessary processing. So, my current understanding of the intention of `show(transform(body))` would be:
* `body` is the value of `response.read()` at this point in the Exercises, which should be a utf8-encoded String
  * Example: `<!doctype><html><head>.*</head><body><p>&lt;html /&gt; is nifty</p></body></html>`
  * If `view-source` is present in the scheme, we pass to `transform(body)`, escaping all wakkas and ampersands
  * This would return the string: `&lt;!doctype&gt;....&lt;p&gt;&amp;lt;html /&amp;gt; is nifty&lt;/p&gt;...&lt;/html&gt;`
* `show` processes the input `body {String}` and outputs content in `<body>` (still omitting HTML tags for Chapter 1 at this Exercise?)
* For the above example, `show(body)` would output `<html /> is nifty` but `show(transform(body))` would output `<!doctype><html><head>.*</head><body><p>&lt;html /&gt; is nifty</p></body></html>`

Is this along the lines of what you mean? I imagine it will make sense too with Chapter 2, I probably refactored too early.

**Reply by mipsumlo** [reply]
Also, `<b>` hurts to still see :( looking forward to CSS support! _really_ great book :)

**Reply by pavpanchekha** [reply]
Yes, you've got it right. Love that you call them wakkas, never heard that before! And, uhh, obscure fact but `<b>` is still a standard, non-deprecated HTML tag. Instead of getting rid of them, the HTML standards folks gave every-so-slightly different meanings to `<strong>` and `<b>`, likewise to `<i>` and so on: https://html.spec.whatwg.org/multipage/text-level-semantics.html#the-b-element

`<b>` is for "text to which attention is being drawn for utilitarian purposes without conveying any extra importance and with no implication of an alternate voice or mood, such as key words in a document abstract, product names in a review, actionable words in interactive text-driven software, or an article lede"

`<strong>` is for "strong importance, seriousness, or urgency"

**Reply by mipsumlo** [reply]
Ah, neat! Thank you for sharing, that's super useful to realize. So web developers just need to be taught this semantic meaning more these days :) And feel free to take the wakka name! I can't take credit for it; but it's one of my favorites.
